### PR TITLE
Restructure derivative handling

### PR DESCRIPTION
**Pull request by JohanWBorg**
Single .derivative attribute instead of splitting ETA/EPS in EstimationStep. Derivatives are always with respect to the prediction.
Derivatives are stored as a tuple of tuples where each tuple represent one derivative. Each element within the tuples are required to be a symbol of type pharmpy.basic.Expr.

### CODE REVIEW

File: src/pharmpy/model/execution_steps.py [hunk 1]
```diff
@@ -153,8 +154,7 @@ def __init__(
         solver_rtol: Optional[int] = None,
         solver_atol: Optional[int] = None,
         tool_options: Optional[frozenmapping[str, Any]] = None,
-        eta_derivatives: Optional[tuple[str, ...]] = None,
-        epsilon_derivatives: Optional[tuple[str, ...]] = None,
+        derivatives: tuple[tuple[Type[Expr]], ...] = (),
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:157]: It should be `Expr` rather than `Type[Expr]`

File: src/pharmpy/model/execution_steps.py [hunk 2]
```diff
@@ -196,8 +195,7 @@ def create(
         solver_rtol: Optional[int] = None,
         solver_atol: Optional[int] = None,
         tool_options: Optional[Mapping[str, Any]] = None,
-        eta_derivatives: Optional[Sequence[str]] = None,
-        epsilon_derivatives: Optional[Sequence[str]] = None,
+        derivatives: tuple[tuple[Type[Expr]], ...] = (),
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:198]: `Expr` rather than `Type[Expr]`

File: src/pharmpy/model/execution_steps.py [hunk 3]
```diff
@@ -272,6 +264,32 @@ def _canonicalize_and_check_method(method: str) -> str:
             )
         return method
 
+    @staticmethod
+    def _canonicalize_derivatives(derivatives: tuple) -> tuple:
+        if isinstance(derivatives, tuple):
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:269]: Hard to follow because of too deep nesting. Check for the problematic case and raise and have no `else`

File: src/pharmpy/model/execution_steps.py [hunk 4]
```diff
@@ -352,14 +370,9 @@ def predictions(self) -> Optional[tuple[str, ...]]:
...
     @property
-    def eta_derivatives(self) -> Optional[tuple[str, ...]]:
-        """List of names of etas for which to calculate derivatives"""
-        return self._eta_derivatives
-
-    @property
-    def epsilon_derivatives(self) -> Optional[tuple[str, ...]]:
-        """List of names of epsilons for which to calculate derivatives"""
-        return self._epsilon_derivatives
+    def derivatives(self) -> tuple[tuple[Type[Expr]], ...]:
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:373]: The same type hint issue

File: tests/model/test_estimation.py [hunk 5]
```diff
@@ -189,13 +193,36 @@ def test_getitem():
 
 
 def test_properties():
-    a = EstimationStep.create('foce', epsilon_derivatives=['EPS(1)'])
-    assert a.epsilon_derivatives == ('EPS(1)',)
-    a = EstimationStep.create('foce', eta_derivatives=['ETA(1)'])
-    assert a.eta_derivatives == ('ETA(1)',)
-    a = EstimationStep.create('foce', predictions=['PRED'])
+    d = tuple(map(Expr.symbol, ('EPS(1)',)))
```
Comment (rikardn) [inline tests/model/test_estimation.py:196]: Pharmpy symbols should not contain parentheses. (parsing becomes ambiguous)

**Review by rikardn — CHANGES_REQUESTED** [review]

File: src/pharmpy/model/execution_steps.py (hunk 1 above)
Comment (JohanWBorg) [reply src/pharmpy/model/execution_steps.py]: Changed all types to `Expr`

File: src/pharmpy/model/execution_steps.py (hunk 3 above)
Comment (JohanWBorg) [reply src/pharmpy/model/execution_steps.py]: Restructured for clarity

File: tests/model/test_estimation.py (hunk 5 above)
Comment (JohanWBorg) [reply tests/model/test_estimation.py]: Switched the symbols to instead use underlines ("ETA_1")

File: src/pharmpy/model/execution_steps.py [hunk 6]
```diff
@@ -272,6 +264,29 @@ def _canonicalize_and_check_method(method: str) -> str:
             )
         return method
 
+    @staticmethod
+    def _canonicalize_derivatives(derivatives: tuple) -> tuple:
+        if not isinstance(derivatives, tuple):
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:269]: This is not what we want. For `create` any `Sequence` should be fine. We simply convert it to `tuple`.

File: src/pharmpy/model/execution_steps.py [hunk 7]
```diff
@@ -272,6 +264,29 @@ def _canonicalize_and_check_method(method: str) -> str:
...
+            raise ValueError(f"Derivatives must be given in a tuple. Got '{type(derivatives)}")
+
+        if not all(isinstance(d, tuple) for d in derivatives):
+            raise ValueError("Each derivative must be given as 'tuple'")
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:273]: With the same logic as above. Cast and only raise if it cannot be converted.

File: src/pharmpy/model/execution_steps.py [hunk 8]
```diff
@@ -272,6 +264,29 @@ def _canonicalize_and_check_method(method: str) -> str:
...
+        for d in derivatives:
+            for d_arg in d:
+                if not (isinstance(d_arg, Expr) and d_arg.is_symbol()):
+                    raise ValueError(
+                        f"Each derivative argument must be a symbol of type 'Expr'. "
+                        f"Found '{d_arg}' of type {type(d_arg)}"
+                    )
+
+        derivatives = tuple(d if isinstance(d, tuple) else (d,) for d in derivatives)
+        # Sorting of symbols creates unpredictable order
```
Comment (rikardn) [inline src/pharmpy/model/execution_steps.py:284]: Strange comment. The order becomes predictable (at least to us).

**Review by rikardn — CHANGES_REQUESTED** [review]

File: src/pharmpy/model/execution_steps.py (hunk 6 above)
Comment (JohanWBorg) [reply src/pharmpy/model/execution_steps.py:269]: Oh ok, i misunderstood this then, will change it to allow any sequence

File: src/pharmpy/model/execution_steps.py (hunk 8 above)
Comment (JohanWBorg) [reply src/pharmpy/model/execution_steps.py:284]: Understandable. The comment is referring to an issue where sorting "symbols" of `Expr` type gave an unpredictable order. For this reason, strings are used instead.

Comment (JohanWBorg) [reply src/pharmpy/model/execution_steps.py]: Yes, that would have been a much more accurate description
